### PR TITLE
[Compute] az vm create: New parameter `--platform-fault-domain` in FLEX VMSS mode

### PR DESCRIPTION
**Pull request by qwordy**
**Description**

Resolve https://github.com/Azure/azure-cli/issues/16346
Resolve https://github.com/Azure/azure-cli/issues/16159

[Compute] vm create: New parameter --platform-fault-domain in FLEX VMSS mode
[Compute] vm create: --patch-mode for Linux VM

**Testing Guide**

```
az vm create -g rg -n vm --image UbuntuLTS --platform-fault-domain 0 --vmss vmss
az vm create -g rg -n vm --image UbuntuLTS --enable-agent --patch-mode AutomaticByPlatform
```

**History Notes**

[Compute] vm create: New parameter --platform-fault-domain in FLEX VMSS mode
[Compute] vm create: --patch-mode for Linux VM

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Compute

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 1]
```diff
@@ -341,10 +341,14 @@ def load_arguments(self, _):
...
         c.argument('enable_auto_update', arg_type=get_three_state_flag(), min_api='2020-06-01',
                    help='Indicate whether Automatic Updates is enabled for the Windows virtual machine')
-        c.argument('patch_mode', arg_type=get_enum_type(self.get_models('InGuestPatchMode')), min_api='2020-06-01',
-                   help='Mode of in-guest patching to IaaS virtual machine. Possible values are: Manual - You  control the application of patches to a virtual machine. You do this by applying patches manually inside the VM. In this mode, automatic updates are disabled; the paramater --enable-auto-update must be false. AutomaticByOS - The virtual machine will automatically be updated by the OS. The parameter --enable-auto-update must be true. AutomaticByPlatform - the virtual machine will automatically updated by the OS. The parameter --enable-agent and --enable-auto-update must be true')
+        c.argument('patch_mode', arg_type=get_enum_type(self.get_models('WindowsVMGuestPatchMode')), min_api='2020-12-01',
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py:344]: why the change?

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 2]
```diff
@@ -341,10 +341,14 @@ def load_arguments(self, _):
...
         c.argument('ssh_key_name', help='Use it as public key in virtual machine. It should be an existing SSH key resource in Azure.')
         c.argument('enable_hotpatching', arg_type=get_three_state_flag(), help='Patch VMs without requiring a reboot. --enable-agent must be set and --patch-mode must be set to AutomaticByPlatform', min_api='2020-12-01')
+        c.argument('platform_fault_domain', min_api='2020-06-01',
+                   help='Specify the scale set logical fault domain into which the virtual machine will be created. By default, the virtual machine will by automatically assigned to a fault domain that best maintains balance across available fault domains. This is applicable only if the virtualMachineScaleSet property of this virtual machine is set. The virtual machine scale set that is referenced, must have platform fault domain count. This property cannot be updated once the virtual machine is created. Fault domain assignment can be viewed in the virtual machine instance view')
+        c.argument('linux_patch_mode', arg_type=get_enum_type(self.get_models('LinuxVMGuestPatchMode')), min_api='2020-12-01',
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py:350]: why have a separate linux_patch_mode along with patch_mode? it's not consistent.

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 3]
```diff
@@ -304,12 +304,19 @@ def _build_os_profile():
...
 
+        # Windows patch settings
         if patch_mode is not None and custom_image_os_type.lower() == 'windows':
             os_profile['windowsConfiguration']['patchSettings'] = {
                 'patchMode': patch_mode,
                 'enableHotpatching': enable_hotpatching
             }
 
+        # Linux patch settings
+        if linux_patch_mode is not None and custom_image_os_type.lower() == 'linux':
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:315]: what if user not using custom_image? None check

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 1 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py:344]: Model name changed.
InGuestPatchMode -> WindowsVMGuestPatchMode

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 2 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: Only --patch-mode now

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py (hunk 3 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py]: If can't infer OS type, user must provide one. OS type is inferred in validator.

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 4]
```diff
@@ -341,10 +341,12 @@ def load_arguments(self, _):
...
         c.argument('ssh_key_name', help='Use it as public key in virtual machine. It should be an existing SSH key resource in Azure.')
         c.argument('enable_hotpatching', arg_type=get_three_state_flag(), help='Patch VMs without requiring a reboot. --enable-agent must be set and --patch-mode must be set to AutomaticByPlatform', min_api='2020-12-01')
+        c.argument('platform_fault_domain', min_api='2020-06-01',
+                   help='Specify the scale set logical fault domain into which the virtual machine will be created. By default, the virtual machine will by automatically assigned to a fault domain that best maintains balance across available fault domains. This is applicable only if the virtualMachineScaleSet property of this virtual machine is set. The virtual machine scale set that is referenced, must have platform fault domain count. This property cannot be updated once the virtual machine is created. Fault domain assignment can be viewed in the virtual machine instance view')
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py:349]:
```suggestion
                   help='Specify the scale set logical fault domain into which the virtual machine will be created. By default, the virtual machine will be automatically assigned to a fault domain that best maintains balance across available fault domains. This is applicable only if the virtualMachineScaleSet property of this virtual machine is set. The virtual machine scale set that is referenced, must have platform fault domain count. This property cannot be updated once the virtual machine is created. Fault domain assignment can be viewed in the virtual machine instance view')
```

**Review by houk-ms — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 5]
```diff
@@ -304,12 +304,19 @@ def _build_os_profile():
         if enable_auto_update is not None and custom_image_os_type.lower() == 'windows':
             os_profile['windowsConfiguration']['enableAutomaticUpdates'] = enable_auto_update
 
+        # Windows patch settings
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:307]: should we add some validation on combination of patch_mode value and os?

Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:307]: This is exactly the disadvantage of merging two parameters into one. The validation was done by Azure CLI framework as long as we define `arg_type=get_enum_type(self.get_models('LinuxVMGuestPatchMode')`.

Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:307]: Added

**Review by yungezz — APPROVED** [review]
